src/aeso_ast_infer_types.erl

```diff
@@ -1049,6 +1131,73 @@ check_typedef(Env, {variant_t, Cons}) ->
 
+infer_type_vars_variance(TypeParams, Cons) ->
+    % args from all type constructors
+    FlatArgs = lists:flatten([Args || {constr_t, _, _, Args} <- Cons]),
+
+    Vs = lists:flatten([infer_type_vars_variance(Arg) || Arg <- FlatArgs]),
+    lists:map(fun({tvar, _, TVar}) ->
+                      S = sets:from_list([Variance || {TV, Variance} <- Vs, TV == TVar]),
+                      IsInvariant     = sets:is_element(invariant, S),
+                      IsCovariant     = sets:is_element(covariant, S),
+                      IsContravariant = sets:is_element(contravariant, S),
+                      IsBivariant     = sets:is_element(bivariant, S),
+                      case {IsInvariant, IsCovariant, IsContravariant, IsBivariant} of
+                          {true,      _,     _,    _} -> invariant;
+                          {false,  true,  true,    _} -> invariant;
+                          {false,     _,     _, true} -> bivariant;
+                          {false,  true, false,    _} -> covariant;
+                          {false, false,  true,    _} -> contravariant;
+                          _ -> invariant
+                      end
+              end, TypeParams).
+
+-spec infer_type_vars_variance(utype()) -> [{name(), variance()}].
+infer_type_vars_variance({app_t, _, Type, Args}) ->
+    Variances = case ets_lookup(type_vars_variance, qname(Type)) of
+                    [{_, Vs}] -> Vs;
+                    _ -> lists:duplicate(length(Args), covariant)
+                end,
+    TypeVarsVariance = [{TVar, Variance}
+                        || {{tvar, _, TVar}, Variance} <- lists:zip(Args, Variances)],
+    TypeVarsVariance;
+infer_type_vars_variance(FT = {fun_t, _, [], [{app_t, _, Type, Args}], Res}) ->
+    Variances = case ets_lookup(type_vars_variance, qname(Type)) of
+                    [{_, Vs}] -> Vs;
+                    _ -> lists:duplicate(length(Args), covariant)
+                end,
+    TypeVarsVariance = [{TVar, Variance}
+                        || {{tvar, _, TVar}, Variance} <- lists:zip(Args, Variances)],
+    FlipVariance = fun({TVar, covariant}) -> {TVar, contravariant};
+                      ({TVar, contravariant}) -> {TVar, covariant}
+                   end,
+    Cur = case arrows_in_type(FT) rem 2 of
+              0 -> TypeVarsVariance;
+              1 -> lists:map(FlipVariance, TypeVarsVariance)
+          end,
+    Cur ++ infer_type_vars_variance(Res);
+infer_type_vars_variance(FT = {fun_t, _, [], [{tvar, _, TVar}], Res}) ->
+    Cur = case arrows_in_type(FT) rem 2 of
+              0 -> {TVar, covariant};
+              1 -> {TVar, contravariant}
+          end,
+    [Cur | infer_type_vars_variance(Res)];
+infer_type_vars_variance({fun_t, _, [], [_Arg], Res}) ->
+    infer_type_vars_variance(Res);
+infer_type_vars_variance({tvar, _, TVar}) ->
+    [{TVar, covariant}];
+infer_type_vars_variance(_) -> [].
+
+arrows_in_type({fun_t, _, [], [_Arg], FRes}) ->
```

Comment:
I think it would be easier and simpler to use `infer_type_vars_variance` to dive into the type and just maintain an additional argument that tells what the current variance is. Then you won't need to count this.

Reply:
Done here 39b9bb66cca29f836f8c36a3569df494dbd19aff